pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/FilterParser.java

```diff
@@ -279,25 +282,108 @@ int safeToInt(Long value) throws FilterStringSyntaxException {
+
+        String digits = parseDigits();
+
+        try {
+            return Integer.parseInt(digits);
+        } catch (NumberFormatException e) {
+            throw new FilterStringSyntaxException("invalid numeric argument at " + digits);
+        }
+    }
+
+    private int parseDataLength() throws Exception {
+        if (((char) filterByteArr[index]) != 's') {
+            throw new FilterStringSyntaxException("data length delimiter 's' expected at " +  index);
+        }
+
+        index++;
+        return parseInt();
+    }
+
+    private int parseInt() throws Exception {
+        if (index == filterByteArr.length) {
+            throw new FilterStringSyntaxException("numeric argument expected at " + index);
+        }
+
+        String digits = parseDigits();
+
+        try {
+            return Integer.parseInt(digits);
+        } catch (NumberFormatException e) {
+            throw new FilterStringSyntaxException("invalid numeric argument " + digits);
+        }
+    }
+
+    private Object convertDataType(byte[] byteData, DataType dataType) throws Exception {
+        String data = new String(byteData);
+        try {
+            switch (dataType) {
+                case BIGINT:
+                    return Long.parseLong(data);
+                case INTEGER:
+                case SMALLINT:
+                    return Integer.parseInt(data);
+                case FLOAT8:
+                case REAL:
+                    return Float.parseFloat(data);
+                case NUMERIC:
+                    return Double.parseDouble(data);
+                case TEXT:
+                case VARCHAR:
+                case BPCHAR:
+                    return data;
+                case BOOLEAN:
+                    return Boolean.parseBoolean(data);
+                case DATE:
+                    return Date.valueOf(data);
+                case TIMESTAMP:
+                    return Timestamp.valueOf(data);
+                default:
+                    throw new FilterStringSyntaxException("DataType " + dataType.toString() + " unsupported");
+            }
+        } catch (NumberFormatException nfe) {
+            throw new FilterStringSyntaxException("failed to parse number data type starting at " + index);
+        }
+    }
     /**
      * Parses either a number or a string.
      */
     private Object parseParameter() throws Exception {
-        if (index == filterString.length()) {
+        if (index == filterByteArr.length) {
             throw new FilterStringSyntaxException("argument should follow at " + index);
         }
 
-        return senseString()
-                ? parseString()
-                : parseNumber();
-    }
+        DataType dataType = DataType.get(parseConstDataType());
+        if (dataType == DataType.UNSUPPORTED_TYPE) {
+            throw new FilterStringSyntaxException("invalid DataType OID at " + (index - 1));
+        }
+
+        int dataLength = parseDataLength();
+        if (dataLength < 1) {
```

Comment:
For the case when TEXT field is checked with = against "", bridge uses s0do5 in the expression string.